homeassistant/components/vicare/number.py

```diff
@@ -0,0 +1,194 @@
+"""Number for ViCare."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from contextlib import suppress
+from dataclasses import dataclass
+import logging
+
+from PyViCare.PyViCareDevice import Device as PyViCareDevice
+from PyViCare.PyViCareDeviceConfig import PyViCareDeviceConfig
+from PyViCare.PyViCareHeatingDevice import (
+    HeatingDeviceWithComponent as PyViCareHeatingDeviceWithComponent,
+)
+from PyViCare.PyViCareUtils import (
+    PyViCareInvalidDataError,
+    PyViCareNotSupportedFeatureError,
+    PyViCareRateLimitError,
+)
+from requests.exceptions import ConnectionError as RequestConnectionError
+
+from homeassistant.components.number import NumberEntity, NumberEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EntityCategory
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import ViCareRequiredKeysMixin
+from .const import DOMAIN, VICARE_API, VICARE_DEVICE_CONFIG
+from .entity import ViCareEntity
+from .utils import is_supported
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class ViCareNumberEntityDescription(NumberEntityDescription, ViCareRequiredKeysMixin):
+    """Describes ViCare number entity."""
+
+    value_setter: Callable[[PyViCareDevice, float], str | None] | None = None
+
+
+CIRCUIT_ENTITY_DESCRIPTIONS: tuple[ViCareNumberEntityDescription, ...] = (
+    ViCareNumberEntityDescription(
+        key="heating curve shift",
+        name="Heating curve shift",
+        icon="mdi:plus-minus-variant",
+        entity_category=EntityCategory.CONFIG,
+        value_getter=lambda api: api.getHeatingCurveShift(),
+        value_setter=lambda api, shift: api.setHeatingCurve(
+            shift, api.getHeatingCurveSlope()
+        ),
+        native_min_value=-13,
+        native_max_value=40,
+        native_step=1,
+    ),
+    ViCareNumberEntityDescription(
+        key="heating curve slope",
+        name="Heating curve slope",
+        icon="mdi:slope-uphill",
+        entity_category=EntityCategory.CONFIG,
+        value_getter=lambda api: api.getHeatingCurveSlope(),
+        value_setter=lambda api, slope: api.setHeatingCurve(
+            api.getHeatingCurveShift(), slope
+        ),
+        native_min_value=0.2,
+        native_max_value=3.5,
+        native_step=0.1,
```

Comment:
These values are picked from my heating device, but the vary a little bit for other devices (min value). But https://github.com/somm15/PyViCare/pull/338 will add the functionality to read this values from the API.

https://github.com/somm15/PyViCare/blob/8ba411483a865e074d1146fd1b8b7a8c4f4be122/tests/response/VitovalorPT2.json#L439-L463C10

https://github.com/somm15/PyViCare/blob/8ba411483a865e074d1146fd1b8b7a8c4f4be122/tests/response/Vitocal300G.json#L335-L350

https://github.com/somm15/PyViCare/blob/8ba411483a865e074d1146fd1b8b7a8c4f4be122/tests/response/Solar.json#L17-L40

https://github.com/somm15/PyViCare/blob/8ba411483a865e074d1146fd1b8b7a8c4f4be122/tests/response/VitolaUniferral.json#L1910-L1933